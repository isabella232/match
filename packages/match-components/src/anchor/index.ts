export * from "./anchor";
export * from "./constants";
export * from "./types";
