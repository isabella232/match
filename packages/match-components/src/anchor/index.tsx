import * as React from "react";
import * as PropTypes from "prop-types";
import { StyledAnchor } from "./styles";
import { AnchorVariant, AnchorProps } from "./types";

export interface ExternalProps {
  rel: string | undefined;
  target: string | undefined;
}

const EXTERNAL_URL_REGEX = /^(https?:)\S*$/;

const secureExternalLink = (href: string | undefined): ExternalProps => {
  if (href) {
    if (EXTERNAL_URL_REGEX.test(href)) {
      return { rel: "noreferrer noopener", target: "_blank" };
    }
  }
  return { rel: undefined, target: undefined };
};

const Anchor: React.FC<AnchorProps> = ({ ...props }) => {
  if (!props.href) console.warn("href must be provided for an anchor");
  const externalURL = secureExternalLink(props.href);
  return (
    <StyledAnchor
      {...props}
      rel={props.rel ? props.rel : externalURL.rel}
      target={props.target ? props.target : externalURL.target}
    />
  );
};

Anchor.propTypes = {
  variant: PropTypes.oneOf(Object.values(AnchorVariant)),
  href: PropTypes.string,
  target: PropTypes.string,
  rel: PropTypes.string,
  noUnderline: PropTypes.bool,
};

Anchor.defaultProps = {
  variant: AnchorVariant.TEXT,
  noUnderline: false,
};

export { Anchor, AnchorVariant, AnchorProps };
