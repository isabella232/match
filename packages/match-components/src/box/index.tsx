import * as React from "react";
import { spaceTokenProp } from "@twilio-labs/match-props";
import { BoxProps } from "./types";
import { StyledBox } from "./styles";

const Box: React.FC<BoxProps> = ({ ...props }) => {
  return <StyledBox {...props} />;
};

Box.propTypes = {
  padding: spaceTokenProp,
  paddingLeft: spaceTokenProp,
  paddingRight: spaceTokenProp,
  paddingBottom: spaceTokenProp,
  paddingTop: spaceTokenProp,
  margin: spaceTokenProp,
  marginLeft: spaceTokenProp,
  marginRight: spaceTokenProp,
  marginBottom: spaceTokenProp,
  marginTop: spaceTokenProp,
};

export { Box, BoxProps };
