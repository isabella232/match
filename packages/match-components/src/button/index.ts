export * from "./constants";
export * from "./button";
export * from "./types";
