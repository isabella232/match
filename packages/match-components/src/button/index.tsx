import * as React from "react";
import * as PropTypes from "prop-types";
import { uid } from "react-uid";
import { StyledButton, StyledPrompt } from "./styles";
import { ButtonVariant, ButtonType, ButtonSize } from "./types";
import type { ButtonProps } from "./types";

const Button: React.FC<ButtonProps> = ({
  type,
  prompt,
  children,
  ...props
}) => {
  if (props.download && !props.href)
    console.warn("[Button]: Href must be provided for a download link.");

  return (
    <StyledButton
      forwardedAs={props.href ? "a" : undefined}
      type={!props.href ? type : undefined}
      {...props}
    >
      {Array.isArray(children)
        ? children.map((child) => {
            if (React.isValidElement(child))
              return React.cloneElement(child, { key: uid(child) });
            if (typeof child === "string") return child.trim();
          })
        : children}
      {prompt && <StyledPrompt />}
    </StyledButton>
  );
};

Button.propTypes = {
  variant: PropTypes.oneOf(Object.values(ButtonVariant)),
  type: PropTypes.oneOf(Object.values(ButtonType)),
  size: PropTypes.oneOf(Object.values(ButtonSize)),
  href: PropTypes.string,
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  download: PropTypes.bool,
  prompt: PropTypes.bool,
};

Button.defaultProps = {
  variant: ButtonVariant.PRIMARY,
  type: ButtonType.BUTTON,
  size: ButtonSize.NORMAL,
  disabled: false,
  fullWidth: false,
  download: false,
  prompt: false,
};

Button.displayName = "Button";

export { Button, ButtonVariant, ButtonType, ButtonSize };
export type { ButtonProps };
