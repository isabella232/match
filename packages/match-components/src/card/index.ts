export * from "./card";
export * from "./constants";
export * from "./types";
