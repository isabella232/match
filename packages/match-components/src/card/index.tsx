import * as React from "react";
import * as PropTypes from "prop-types";
import { paddingPropTypes, marginPropTypes } from "@twilio-labs/match-props";
import { CardVariant } from "./types";
import type { CardProps } from "./types";
import { StyledCard } from "./styles";

const Card: React.FC<CardProps> = (props) => <StyledCard {...props} />;

Card.propTypes = {
  ...paddingPropTypes,
  ...marginPropTypes,
  variant: PropTypes.oneOf(Object.values(CardVariant)),
};

Card.defaultProps = {
  variant: CardVariant.PRIMARY,
};

Card.displayName = "Card";

export { Card, CardVariant };
export type { CardProps };
