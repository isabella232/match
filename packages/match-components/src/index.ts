export * from "./anchor";
export * from "./button";
export * from "./card";
export * from "./code";
export * from "./form";
export * from "./grid";
export * from "./heading";
export * from "./input";
export * from "./paragraph";
export * from "./richtext";
export * from "./separator";
export * from "./snippet";
export * from "./textarea";
export * from "./truncate";
export * from "./visually-hidden";
