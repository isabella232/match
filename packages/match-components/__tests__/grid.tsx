import "jest-styled-components";
import * as React from "react";
import { render } from "@testing-library/react";
import { withTheme } from "@twilio-labs/match-themes";
import { Grid, Cell } from "../src";

const GridWithTheme = withTheme()(Grid);

describe("Spaces", () => {
  test("snapshot", () => {
    const { container } = render(
      <GridWithTheme
        gridTemplateColumns="1fr 2fr"
        gridTemplateAreas="sidebar main"
        gridGap="scale100"
      >
        <Cell gridArea="sidebar" />
        <Cell gridArea="main" />
      </GridWithTheme>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
