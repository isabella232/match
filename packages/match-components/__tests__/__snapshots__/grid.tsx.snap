// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spaces snapshot 1`] = `
.c0 {
  color: #1F304C;
  font-weight: 400;
  font-size: 1rem;
  font-family: Whitney SSm A,Whitney SSm B,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c1 {
  display: grid;
  grid-template-columns: 1fr 2fr;
  grid-template-areas: sidebar main;
  grid-gap: 1rem;
}

.c2 {
  grid-area: sidebar;
}

.c3 {
  grid-area: main;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    />
    <div
      class="c3"
    />
  </div>
</div>
`;
